### ISSUE TITLE
ScatterView is slow in certain cases

### ISSUE THREAD

**Issue by ibaned** [kokkos/kokkos#1390]
This is a placeholder to track certain performance issues with `ScatterView`. Data is currently internal only, but this issue will at least track whether it was resolved.

**Comment by stanmoore1** [comment]
I think it is a layout issue with the reduction operation. If I use 2D layout right, the time spent in reduction is 6.1 seconds. If I collapse the view, i.e. make it 1D but keep the same span by making it larger in the first dimension, the time spent in reduction is 1.9 seconds, even though the work should be the same. I’m guessing the way we are iterating over the view in the reduction is thrashing cache or something.

**Comment by stanmoore1** [comment]
Scratch that, my view was resized in another place back to the smaller value without my notice. So the time spent in the kernel is nearly the same for 1D vs 2D. Now I'm really perplexed.

**Comment by ibaned** [comment]
Note that one really shouldn't use a persistent `ScatterView` if the original view resizes.

**Comment by stanmoore1** [comment]
True--in this case I was using temporary ScatterViews so they are reallocated before every use.

**Comment by crtrott** [comment]
In the case there I want to reallocate occasionally it would be great if the scatter view would track what its dimensions are, because I can't check whether the original one changed. Or maybe a function which returns the original view, or something like that.

**Comment by stanmoore1** [comment]
Another common use case is that a View is allocated with some extra space, and it would improve performance if I could specify the limits of the reduction so it wasn't over that extra space.

**Comment by stanmoore1** [comment]
@ibaned @crtrott The root cause of the slowdown in my case was that the ScatterView was allocated with a lot of extra space, which made the reduction really expensive. So I think this particular issue can be closed.

**Comment by ibaned** [comment]
Also to answer @crtrott's question, I think `scatter_view.subview()` turns a subview whose extents should match the original.

If @crtrott confirms that the global versus local sizing was all there was to this, then we'll close it.

**Comment by ibaned** [comment]
@stanmoore1 for the "original with extra space use", it may be as simple as using an appropriately small subview of the original in order to initialize the `ScatterView`.

**Comment by stanmoore1** [comment]
That is a good idea.

**Comment by stanmoore1** [comment]
Though that wouldn't really work for the persistent case.

**Comment by stanmoore1** [comment]
>In the case there I want to reallocate occasionally it would be great if the scatter view would track what its dimensions are, because I can't check whether the original one changed. Or maybe a function which returns the original view, or something like that.

ScatterView has a `resize` method. That is what I use for the persistent case to reallocate.

**Comment by crtrott** [comment]
This is not everything. The reduction kernel should either run in N steps to avoid jumping back and forth in memory or better use some tiling strategy. Right now it does 
```c++
parallel_for(L, KOKKOS_LAMBDA (const int& i) {
   for(int j=0; j<n; j++)
      dat[i] += src[i+stride*j];
});
```
We need this to be something like:
```c++
for(int j=0; j<n; j++)
  parallel_for(L, KOKKOS_LAMBDA (const int& i) {
      dat[i] += src[i+stride*j];
  });
```

Or better some hierarchical tiling based algorithm:

```c++
  parallel_for(TeamPolicy<>(L/T,CUDA?AUTO:1), KOKKOS_LAMBDA (const auto& team) {
    for(int j=0; j<n; j++) 
      parallel_for(TeamThreadRange(team, team.league_rank()*T, (team.league_rank()+1)*T), 
        [&] (const int& i) {
          dat[i] += src[i+stride*j];
      });
  });
```
